Action View Form Helpers > 3. Making Select Boxes with Ease (#320)
campuscode/rails-guides-pt-BR: pt-BR/form_helpers.md

IMPORTANT: All forms using `form_with` implement `remote: true` by default. These forms will submit data using an XHR (Ajax) request. To disable this include `local: true`. To dive deeper see [Working with JavaScript in Rails](working_with_javascript_in_rails.html#remote-elements) guide.

> Fazendo (*Select boxes*) caixas de seleção com facilidade

Reviewer: ```suggestion
Fazendo Caixas de Seleção (*Select Boxes*) com Facilidade
```

Acho que aqui podia inverter a ordem pra ficar mais fluido a leitura, o que acha?

Coloquei algumas letras maiúsculas também pra ficar como titulo o que acha?

Author: Achei que realmente que melhorou a fluência, irei corrigir
Muito obrigado!

-----------------------------

As caixas de seleção em HTML requerem uma quantidade significativa de marcação (um elemento `OPTION` para cada opção de escolha), portanto, faz mais sentido que sejam geradas dinamicamente. 

Esta é a aparência da marcação:
```html
<select name="city_id" id="city_id">
  <option value="1">Lisboa</option>
  <option value="2">Madri</option>
  <option value="3">Berlim</option>
</select>

> ```

Reviewer 2: Eu manteria em inglês neste caso pra manter consistente com as outras partes que não traduzimos, nesse caso acho que não fica muito estranho mas em alguns casos pode causar dúvidas a tradução.

> ```

Reviewer: Eu achei legal o fato de ser em pt-BR porque parece com algo que eu esperaria lendo a doc. Mas a consistência de páginas realmente fica problemática :/

> ```

Author: Irei corrigir, para manter o padrão. Obrigado, pessoal

Aqui você tem uma lista de cidades cujos nomes são apresentados ao usuário. Internamente, o aplicativo deseja apenas manipular seus IDs, para que sejam usados como o atributo de valor das opções. Vamos ver como o Rails pode ajudar aqui.

> ### As tags de seleção (*Select*) e opção (*Option*)

Reviewer: ```suggestion
### As Tags de Seleção (*Select*) e Opção (*Option*)
```

Acho que letras em maiúscula dá um pouco de cara de titulo, o que acha?

> O auxiliar mais genérico é `select_tag`, que - como o nome indica - simplesmente gera a tag `SELECT` que encapsula uma string de opções:

Reviewer: ```suggestion
O auxiliar mais genérico é `select_tag`, que - como o nome indica - simplesmente gera a tag `SELECT` que encapsula uma *string* de opções:
```

Acho que string poderia ficar em itálico por ser termo "estrangeiro" o que acha?

```erb
<%= select_tag(:city_id, raw('<option value="1">Lisboa</option><option value="2">Madri</option><option value="3">Berlim</option>')) %>
```

> Isso é um começo, porém não cria dinamicamente as tags de opção. Você pode gerar tags de opção com o auxiliar `options_for_select`:

Reviewer 3: ```suggestion
Isso é um começo, porém não cria dinamicamente as tags de opção. Você pode gerar tags de opção com o *helper* `options_for_select`:
```

Seguindo o [manual de tradução](https://github.com/campuscode/rails-guides-pt-BR/blob/master/TRANSLATION_MANUAL.md#gloss%C3%A1rio), é preferível não traduzir o termo *helper*. O que acha de deixá-los apenas em itálico mesmo?

```html+erb
<%= options_for_select([['Lisboa', 1], ['Madri', 2], ['Berlim', 3]]) %>
```

Saída:

```html
<option value="1">Lisboa</option>
<option value="2">Madri</option>
<option value="3">Berlim</option>
```

> O primeiro argumento para `options_for_select` é um *array* aninhado onde cada elemento tem dois elementos: texto da opção (nome da cidade) e valor da opção (id da cidade). O valor da opção é o que será enviado ao seu *controller*. Freqüentemente, esse será o id de um objeto de banco de dados correspondente, mas não precisa ser o caso.

Reviewer: ```suggestion
O primeiro argumento para `options_for_select` é um *array* aninhado onde cada elemento tem dois elementos: texto da opção (nome da cidade) e valor da opção (id da cidade). O valor da opção é o que será enviado ao seu *controller*. Frequentemente, esse será o id de um objeto de banco de dados correspondente, mas não precisa ser o caso.
```

Acho que com a mudança no acordo ortográfico a trema saiu do português, pode dar uma olhada?

Sabendo disso, é possível combinar `select_tag` e `options_for_select` para obter a marcação completa desejada:

```erb
<%= select_tag(:city_id, options_for_select(...)) %>
```

`options_for_select` permite pré-selecionar uma opção passando seu valor.

```html+erb
<%= options_for_select([['Lisboa', 1], ['Madri', 2], ['Berlim', 3]], 2) %>
```

Saída:

```html
<option value="1">Lisboa</option>
<option value="2" selected="selected">Madri</option>
<option value="3">Berlim</option>
```

Sempre que o Rails vê que o valor interno de uma opção sendo gerada corresponde a este valor, ele adicionará o atributo `selected` aquela opção.

É possível adicionar atributos arbitrários às opções usando hashes:

```html+erb
<%= options_for_select(
  [
    ['Lisboa', 1, { 'data-size' => '2.8 million' }],
    ['Madri', 2, { 'data-size' => '3.2 million' }],
    ['Berlim', 3, { 'data-size' => '3.4 million' }]
  ], 2
) %>
```

Saída:

```html
<option value="1" data-size="2.8 million">Lisboa</option>
<option value="2" selected="selected" data-size="3.2 million">Madri</option>
<option value="3" data-size="3.4 million">Berlim</option>
```

> ### *Select Boxes* Caixas de seleção para lidar com objetos de modelo

Reviewer: ```suggestion
### Caixas de Seleção (*Select Boxes*) com Objetos *Model*
```

Não sei se ficou muito bom mas achei um pouco confuso a parte do "lidar" até o inglês soa um pouco estranho o que acha?

Não costumamos traduzir o termo **model** só colocar em itálico por ser um termo técnico. [No manual de tradução](https://github.com/campuscode/rails-guides-pt-BR/blob/master/TRANSLATION_MANUAL.md#gloss%C3%A1rio) tem alguns outros termos também, se quiser dar uma olhada.

> Na maioria dos casos, os controles de formulário serão vinculados a um *model* específico e, como você pode esperar, o Rails fornece helpers personalizados para esse propósito. Consistente com outros *helpers* de formulário, ao lidar com um objeto de *model* elimine o sufixo `_tag` de `select_tag`:

Reviewer 3: ```suggestion
Na maioria dos casos, os controles de formulário serão vinculados a um *model* específico e, como você pode esperar, o Rails fornece *helpers* personalizados para esse propósito. Consistente com outros *helpers* de formulário, ao lidar com um objeto de *model* elimine o sufixo `_tag` de `select_tag`:
```

> Na maioria dos casos, os controles de formulário serão vinculados a um modelo específico e, como você pode esperar, o Rails fornece helpers personalizados para esse propósito. Consistente com outros *helpers* de formulário, ao lidar com um objeto de modelo elimine o sufixo `_tag` de` select_tag`:

Reviewer: ```suggestion
Na maioria dos casos, os controles de formulário serão vinculados a um *model* específico e, como você pode esperar, o Rails fornece helpers personalizados para esse propósito. Consistente com outros *helpers* de formulário, ao lidar com um objeto de *model* elimine o sufixo `_tag` de` select_tag`:
```

Se o *controller* definiu `@person` e o city_id dessa pessoa é 2:

```ruby
@person = Person.new(city_id: 2)
```

```erb
<%= select(:person, :city_id, [['Lisboa', 1], ['Madri', 2], ['Berlim', 3]]) %>
```

produz uma saída semelhante a

```html
<select name="person[city_id]" id="person_city_id">
  <option value="1">Lisboa</option>
  <option value="2" selected="selected">Madri</option>
  <option value="3">Berlim</option>
</select>
```

Observe que o terceiro parâmetro, o *array* de opções, é o mesmo tipo de argumento que você passa para `options_for_select`. Uma vantagem aqui é que você não precisa se preocupar em pré-selecionar a cidade correta se o usuário já tiver uma - o Rails fará isso para você lendo o atributo `@person.city_id`.

> Tal como acontece com outros auxiliares, se você fosse usar o *helper* `select` em um construtor de formulário com escopo para o objeto `@person`, a sintaxe seria:

Reviewer 3: ```suggestion
Tal como acontece com outros * helpers*, se você fosse usar o *helper* `select` em um construtor de formulário com escopo para o objeto `@person`, a sintaxe seria:
```

```erb
<%= form_with model: @person do |person_form| %>
  <%= person_form.select(:city_id, [['Lisboa', 1], ['Madri', 2], ['Berlim', 3]]) %>
<% end %>
```

Você também pode passar um bloco para o *helper* `select`:

```erb
<%= form_with model: @person do |person_form| %>
  <%= person_form.select(:city_id) do %>
    <% [['Lisboa', 1], ['Madri', 2], ['Berlim', 3]].each do |c| %>
      <%= content_tag(:option, c.first, value: c.last) %>
    <% end %>
  <% end %>
<% end %>
```

> WARNING: Se você estiver usando `select` ou ajudantes semelhantes para definir uma associação `belongs_to`, você deve passar o nome da chave estrangeira (no exemplo acima `city_id`), não o nome da própria associação.

Reviewer: ```suggestion
WARNING: Se você estiver usando `select` ou *helpers* semelhantes para definir uma associação `belongs_to`, você deve passar o nome da chave estrangeira (no exemplo acima `city_id`), não o nome da própria associação.
```

Mesmo caso do de cima, pode dar uma olhada?

WARNING: Quando `:include_blank` ou`:prompt` não estão presentes, `: include_blank` é forçado a *true* se o atributo de seleção `required` for *true*, display `size` é um e` multiple` não é *true*.

> ### Tags de opção (*options*) de uma coleção de objetos arbitrários

Reviewer: ```suggestion
### Tags de Opção (*Option Tags*) de uma Coleção de Objetos Arbitrários
```

Uma sugestão parecida com algumas acima, só deixando tudo em maiúsculo de invertendo a ordem para pt-BR (en) pode dar uma olhada?

Gerar tags de opções com `options_for_select` requer que você crie um *array* contendo o texto e valor para cada opção. Mas e se você tivesse um *model* `City` (talvez um Active Record) e quisesse gerar tags de opção de uma coleção desses objetos? Uma solução seria fazer uma matriz aninhada iterando sobre eles:

```erb
<% cities_array = City.all.map { |city| [city.name, city.id] } %>
<%= options_for_select(cities_array) %>
```

> Esta é uma solução perfeitamente válida, entretanto Rails fornece uma alternativa menos prolixa: `options_from_collection_for_select`. Este auxiliar espera uma coleção de objetos arbitrários e dois argumentos adicionais: os nomes dos métodos para ler a opção **value** e **text**, respectivamente:

Reviewer 3: ```suggestion
Esta é uma solução perfeitamente válida, entretanto Rails fornece uma alternativa menos verbosa: `options_from_collection_for_select`. Este *helper* espera uma coleção de objetos arbitrários e dois argumentos adicionais: os nomes dos métodos para ler a opção **value** e **text**, respectivamente:
```

Aqui acredito que podemos manter o termo "verbosa" mesmo, pois é o que usualmente utilizamos no dia a dia. O que você acha?

```erb
<%= options_from_collection_for_select(City.all, :id, :name) %>
```

> Como o nome indica, isso só gera tags de opção. Para gerar uma *select box* funcional, você precisará usar `collection_select`:

Reviewer 3: ```suggestion
Como o nome indica, isso só gera *tags* de opção. Para gerar uma *select box* funcional, você precisará usar `collection_select`:
```

```erb
<%= collection_select(:person, :city_id, City.all, :id, :name) %>
```

Como com outros *helpers*, se você fosse usar o *helper* `collection_select` em um construtor de formulário com escopo para o objeto `@person`, a sintaxe seria:

```erb
<%= form_with model: @person do |person_form| %>
  <%= person_form.collection_select(:city_id, City.all, :id, :name) %>
<% end %>
```

> NOTE: Pares passados para `options_for_select` devem ter o texto primeiro e o valor depois, entretanto com `options_from_collection_for_select` devem ter o método do valor primeiro e o método do texto depois.

Reviewer 3: ```suggestion
NOTE: Pares passados para `options_for_select` devem ter o texto primeiro e o valor depois, entretanto, com `options_from_collection_for_select` devem ter o método do valor primeiro e o método do texto depois.
```

> ### Fuso horário e seleção de país

Reviewer: ```suggestion
### Fuso horário e Seleção de País
```

> Para usar o suporte de fuso horário no Rails, você tem que perguntar aos seus usuários em que fuso horário eles estão. Fazer isso exigiria a geração de opções selecionadas de uma lista de *[`ActiveSupport::TimeZone`](https://api.rubyonrails.org/classes/ActiveSupport/TimeZone.html)* objetos usando `collection_select`, mas você pode simplesmente usar o helper` time_zone_select` que já envolve isto:

Reviewer: ```suggestion
Para usar o suporte de fuso horário no Rails, você tem que perguntar aos seus usuários em que fuso horário eles estão. Fazer isso exigiria a geração de opções selecionadas de uma lista de *[`ActiveSupport::TimeZone`](https://api.rubyonrails.org/classes/ActiveSupport/TimeZone.html)* objetos usando `collection_select`, mas você pode simplesmente usar o *helper* `time_zone_select` que já envolve isto:
```

```erb
<%= time_zone_select(:person, :time_zone) %>
```

Existe também o *helper* `time_zone_options_for_select` para uma maneira mais manual (portanto mais customizável) de fazer isso. Leia a [documentação da API](https://api.rubyonrails.org/classes/ActionView/Helpers/FormOptionsHelper.html#method-i-time_zone_options_for_select) para aprender sobre os possíveis argumentos para esses dois métodos.

> Rails _tinha_ um *helper* `country_select` para escolher os países, mas isso foi extraído para o [plugin country_select](https://github.com/stefanpenner/country_select).

Reviewer 3: ```suggestion
O Rails tinha um *helper* `country_select` para escolher os países, mas foi extraído para o [plugin country_select](https://github.com/stefanpenner/country_select).
```
